File: lib/python/opentoken/src/main/opentoken/attributes/person/sex_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation.regex_validator import RegexValidator


class SexAttribute(BaseAttribute):
    """
    Represents an assigned sex of a person.

    This class extends BaseAttribute and provides functionality for working with
    these type of fields. It recognizes "Sex" or "Gender" as valid aliases for
    this attribute type.

    The attribute performs normalization on input values, converting them to a
    standard format (Male or Female).
    """

    NAME = "Sex"
    ALIASES = [NAME, "Gender"]

    # Regular expression pattern for validating sex/gender values
    VALIDATE_REGEX = r"^\s*([Mm](ale)?|[Ff](emale)?)\s*$"

    def __init__(self):
        validation_rules = [
            RegexValidator(self.VALIDATE_REGEX)
        ]
        super().__init__(validation_rules)

    def get_name(self) -> str:
        return self.NAME

    def get_aliases(self) -> List[str]:
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize sex value to 'Male' or 'Female'."""
        if not value:
            return None

        first_char = value.strip()[0].lower()
        if first_char == 'm':
            return "Male"
        elif first_char == 'f':
            return "Female"
        else:
            return None  # Return None if can't normalize

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/person/SexAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import java.util.List;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.RegexValidator;

/**
 * Represents an assigned sex of a person.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * these type of fields. It recognizes "Sex" or "Gender" as valid aliases for
 * this attribute type.
 * 
 * The attribute performs normalization on input values, converting them to a
 * standard format (M or F).
 */
public class SexAttribute extends BaseAttribute {

    private static final String NAME = "Sex";
    private static final String[] ALIASES = new String[] { NAME, "Gender" };

    /**
     * Regular expression pattern for validating sex/gender values.
     *
     * This pattern matches the following formats (case-insensitive):
     *  - "M" or "F"
     *  - "Male" or "Female"
     *
     * Breakdown of the regex:
     *   ^                 Start of string
     *   (                 Start of group:
     *     [Mm](ale)?      'M' or 'm', optionally followed by 'ale'
     *     |               OR
     *     [Ff](emale)?    'F' or 'f', optionally followed by 'emale'
     *   )
     *   $                 End of string
     */
    private static final String VALIDATE_REGEX = "^([Mm](ale)?|[Ff](emale)?)$";

    public SexAttribute() {
        super(
                List.of(
                        new RegexValidator(VALIDATE_REGEX)));
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String normalize(String value) {
        switch (value.charAt(0)) {
            case 'M':
            case 'm':
                return "Male";
            case 'F':
            case 'f':
                return "Female";
            default:
                return null;
        }
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }
}
